### PR TITLE
Warn when using deprecated listeners as loggers

### LINKED ISSUE #23360: When configuration cache is enabled report usages of `Gradle.useLogger()`
`Gradle.useLogger()` is an alternative to `Gradle.useListener()` that can be used to register build listeners. However `useLogger()` does not raise any configuration cache problems when used with an unsupported listener type, whereas `useListener()` does.

### Expected Behavior

When configuration cache is enabled, raise a problem when `Gradle.useLogger()` is called with an unsupported listener type.

### Current Behavior

Unsupported listener types are silently ignored.

### PR DESCRIPTION
**Pull request by mlopatkin**
Fixes #23360

### CODE REVIEW

**Comment by mlopatkin** [comment]
@bot-gradle test BD, ACC

**Review by hythloda — APPROVED** [review]

**Comment by mlopatkin** [comment]
@bot-gradle test ACC

File: subprojects/core-api/src/main/java/org/gradle/BuildListener.java [hunk 1]
```diff
@@ -26,7 +27,7 @@
  * @see org.gradle.api.invocation.Gradle#addListener(Object)
  */
 @EventScope(Scopes.Build.class)
-public interface BuildListener {
+public interface BuildListener extends ListenerDeprecatedInGradleScope {
```
Comment (adammurdoch) [inline subprojects/core-api/src/main/java/org/gradle/BuildListener.java:30]: I wonder if an annotation would be better.

**Review by adammurdoch — APPROVED** [review]

File: subprojects/core-api/src/main/java/org/gradle/BuildListener.java (hunk 1 above)
Comment (mlopatkin) [reply subprojects/core-api/src/main/java/org/gradle/BuildListener.java]: Replaced with an annotation, it has benefit of being internal.

**Comment by mlopatkin** [comment]
@bot-gradle test and merge

**Comment by mlopatkin** [comment]
@bot-gradle test and merge

**Pull request merged**